### ISSUE TITLE
Scrolling breaking semantic highlighting 

### ISSUE THREAD

**Issue by RedCMD** [microsoft/vscode#210295]
Scrolling up and down causes semantic highlighting to randomly stop working

First time it's happened to me
I dare not restart anything atm

other files not affected
its not locked to any certain scroll position (cut off point**s** randomly shifts)

doesn't change 'while' scrolling, only after I've stopped scrolling for ~1sec

Some how it still has its semantic token info applied
![image](https://github.com/microsoft/vscode/assets/33529441/f5565021-e702-4b38-8c70-d427b18ff13e)

But isn't affecting '`dead`' code (darker blue). only `consts`
![image](https://github.com/microsoft/vscode/assets/33529441/5f957be8-0b5c-4ea0-a34c-70f9ce1973d0)
![image](https://github.com/microsoft/vscode/assets/33529441/1d4b0479-db88-4018-86ea-dea833f1fd29)

![image](https://github.com/microsoft/vscode/assets/33529441/14e33eb3-68cb-4ebb-80d0-4ce348574a95)
```
log.ts:419  INFO Started local extension host with pid 7580.
log.ts:429  WARN [cmake-tools]: Couldn't find message for key cmake-tools.configuration.cmake.options.advanced.statusBarVisibility.variant.description.
log.ts:439   ERR Extension 'ms-toolsai.jupyter' wants API proposal 'notebookControllerKind' but that proposal DOES NOT EXIST. Likely, the proposal has been finalized (check 'vscode.d.ts') or was abandoned.
log.ts:419  INFO Extension host (LocalProcess pid: 7580) is unresponsive.
log.ts:419  INFO Extension host (LocalProcess pid: 7580) is responsive.
log.ts:419  INFO [perf] Render performance baseline is 60ms
log.ts:439   ERR Cancelled: Canceled
    at _.B (vscode-file://vscode-app/e:/VSCode-win32-x64/resources/app/out/vs/workbench/workbench.desktop.main.js:785:28974)
    at async _.z (vscode-file://vscode-app/e:/VSCode-win32-x64/resources/app/out/vs/workbench/workbench.desktop.main.js:785:25774)
    at async _.getExtensions (vscode-file://vscode-app/e:/VSCode-win32-x64/resources/app/out/vs/workbench/workbench.desktop.main.js:785:22432)
    at async oe.getExtensions (vscode-file://vscode-app/e:/VSCode-win32-x64/resources/app/out/vs/workbench/workbench.desktop.main.js:2909:13776)
    at async Me.Lc (vscode-file://vscode-app/e:/VSCode-win32-x64/resources/app/out/vs/workbench/workbench.desktop.main.js:2845:25206)
    at async Me.Uc (vscode-file://vscode-app/e:/VSCode-win32-x64/resources/app/out/vs/workbench/workbench.desktop.main.js:2845:27892)
    at async Me.rc (vscode-file://vscode-app/e:/VSCode-win32-x64/resources/app/out/vs/workbench/workbench.desktop.main.js:2845:14713)
    at async vscode-file://vscode-app/e:/VSCode-win32-x64/resources/app/out/vs/workbench/workbench.desktop.main.js:2845:13012
log.ts:429  WARN Settings pattern "application.*" doesn't match any settings
```

```json
  {
	"window.zoomLevel": 1,
	"files.autoSave": "afterDelay",
	"editor.acceptSuggestionOnEnter": "off",
	"window.customMenuBarAltFocus": false,
	"window.enableMenuBarMnemonics": false,
	"files.hotExit": "onExitAndWindowClose",
	"editor.suggestSelection": "recentlyUsedByPrefix",
	"explorer.autoReveal": false,
	"editor.renderControlCharacters": true,
	"[.asm]": {
		"editor.quickSuggestions": {
			"comments": "off",
			"strings": "off",
			"other": "off"
		},
		"editor.parameterHints.enabled": false,
		"editor.formatOnType": false,
		"editor.wordWrap": "off",
		"C_Cpp.default.systemIncludePath": [
			"D:\\Documents\\CASIO\\fx-9860G_SDK\\OS"
		]
	},
	"[.src]": {
		"editor.quickSuggestions": {
			"comments": "off",
			"strings": "off",
			"other": "off"
		},
		"editor.parameterHints.enabled": false,
		"editor.formatOnType": false,
		"editor.wordWrap": "off"
	},
	"[xml]": {
		"editor.wordWrap": "on"
	},
	"files.associations": {
		"*.wasm": "wasm"
	},
	"editor.formatOnType": true,
	"editor.tabSize": 4,
	"editor.detectIndentation": false,
	"editor.insertSpaces": false,
	"editor.useTabStops": false,
	"workbench.editorAssociations": {
		"*.abs": "default",
		"*.wasm": "default"
	},
	"editor.minimap.maxColumn": 10000,
	"diffEditor.renderSideBySide": true,
	"workbench.editor.highlightModifiedTabs": true,
	"editor.largeFileOptimizations": false,
	"workbench.startupEditor": "none",
	"files.maxMemoryForLargeFilesMB": 3072,
	"files.participants.timeout": 10000,
	"workbench.editor.tabCloseButton": "off",
	"workbench.editor.tabSizing": "shrink",
	"workbench.editor.titleScrollbarSizing": "large",
	"editor.selectionHighlight": true,
	"editor.occurrencesHighlight": "singleFile",
	"editor.foldingHighlight": true,
	"editor.renderLineHighlight": "all",
	"workbench.editor.wrapTabs": true,
	"editor.wordWrapColumn": 180,
	"debug.internalConsoleOptions": "neverOpen",
	"editor.stickyTabStops": true,
	"diffEditor.renderIndicators": false,
	"explorer.confirmDragAndDrop": false,
	"markdown.preview.fontSize": 13,
	"security.workspace.trust.untrustedFiles": "open",
	"explorer.confirmDelete": false,
	"git.enableSmartCommit": false,
	"git.autofetch": true,
	"files.exclude": {
		"**/.git": false,
		"desktop.ini": true
	},
	"git.confirmSync": false,
	"json.maxItemsComputed": 1000000,
	"search.maxResults": 50000,
	"diffEditor.maxComputationTime": 30000,
	"editor.trimAutoWhitespace": false,
	"editor.unicodeHighlight.allowedCharacters": {
		"ỿ": true,
		"ſ": true
	},
	"editor.glyphMargin": true,
	"codelens-sample.enableCodeLens": true,
	"editor.unfoldOnClickAfterEndOfLine": true,
	"redhat.telemetry.enabled": false,
	"search.useGlobalIgnoreFiles": true,
	"editor.suggest.insertMode": "replace",
	"javascript.suggest.completeFunctionCalls": true,
	"js/ts.implicitProjectConfig.strictFunctionTypes": false,
	"git.inputValidationSubjectLength": 75,
	"git.ignoreMissingGitWarning": true,
	"editor.guides.highlightActiveIndentation": true,
	"editor.guides.bracketPairs": "active",
	"extensions.autoUpdate": false,
	"editor.renderWhitespace": "boundary",
	"update.mode": "manual",
	"editor.minimap.size": "fill",
	"editor.wrappingIndent": "none",
	"[json]": {
		"editor.defaultFormatter": "vscode.json-language-features"
	},
	"[regex-tmLanguage]": {
		"editor.maxTokenizationLineLength": 500000
	},
	"[javascript]": {
		"editor.defaultFormatter": "vscode.typescript-language-features",
		"editor.maxTokenizationLineLength": 5000
	},
	"[jsonc]": {
		"editor.defaultFormatter": "vscode.json-language-features"
	},
	"powershell.integratedConsole.showOnStartup": false,
	"powershell.integratedConsole.focusConsoleOnExecute": false,
	"powershell.integratedConsole.startInBackground": true,
	"editor.foldingMaximumRegions": 10000,
	"SAS.connectionProfiles": {},
	"workbench.editor.empty.hint": "hidden",
	"editor.unicodeHighlight.invisibleCharacters": false,
	"git.openRepositoryInParentFolders": "never",
	"editor.maxTokenizationLineLength": 500000,
	"workbench.editor.tabActionCloseVisibility": false,
	"files.readonlyExclude": {
		"**/node_modules/vscode-textmate/**/*": true
	},
	"C_Cpp.formatting": "vcFormat",
	"C_Cpp.vcFormat.newLine.beforeOpenBrace.function": "sameLine",
	"C_Cpp.vcFormat.newLine.beforeOpenBrace.block": "sameLine",
	"C_Cpp.vcFormat.indent.caseLabels": true,
	"workbench.editor.enablePreview": true,
	"workbench.editor.closeEmptyGroups": false,
	"[leojsbody.plain.wrap][leojsbody.python.wrap][leojsbody.javascript.wrap][leojsbody.typescript.wrap][leojsbody.c.wrap][leojsbody.cpp.wrap][leojsbody.css.wrap][leojsbody.fortran.wrap][leojsbody.fortran90.wrap][leojsbody.html.wrap][leojsbody.java.wrap][leojsbody.json.wrap][leojsbody.markdown.wrap][leojsbody.php.wrap][leojsbody.restructuredtext.wrap][leojsbody.rust.wrap]": {
		"editor.wordWrap": "on"
	},
	"window.confirmBeforeClose": "never",
	"debug.onTaskErrors": "debugAnyway",
	"workbench.tree.stickyScrollMaxItemCount": 8,
	"editor.stickyScroll.maxLineCount": 7,
	"C_Cpp.codeAnalysis.exclude": {},
	"[json-textmate]": {
		"editor.defaultFormatter": "RedCMD.tmlanguage-syntax-highlighter"
	},
	"javascript.format.semicolons": "insert",
	"javascript.referencesCodeLens.enabled": true,
	"js/ts.implicitProjectConfig.module": "ES2022",
	"js/ts.implicitProjectConfig.target": "ES2022",
	"typescript.disableAutomaticTypeAcquisition": true,
	"typescript.format.semicolons": "insert",
	"typescript.inlayHints.functionLikeReturnTypes.enabled": true,
	"typescript.inlayHints.parameterTypes.enabled": true,
	"typescript.suggest.completeFunctionCalls": true,
	"diffEditor.ignoreTrimWhitespace": false,
	"diffEditor.hideUnchangedRegions.enabled": true
}
```

possibly related: https://github.com/microsoft/vscode/issues/197459

- VS Code Version: 1.88.0
- OS Version: Windows 11

**Comment by VSCodeTriageBot** [comment]
Thanks for creating this issue! It looks like you may be using an old version of VS Code, the latest stable release is 1.88.1. Please try upgrading to the latest version and checking whether this issue remains.

Happy Coding!

**Comment by RedCMD** [comment]
I do think the bot shouldn't be so sensitive

**Comment by RedCMD** [comment]
@hediet 
@alexdima seems to be away

**Comment by alexdima** [comment]
This is concerning, I see it being broken in your video. I wonder if the collapsed code blocks play a part here, in this case we make multiple requests to the TS language service, with different ranges here https://github.com/microsoft/vscode/blob/4af0c6a8f4e8b7ba223d27acd2d894a76a912a5e/src/vs/editor/contrib/semanticTokens/browser/viewportSemanticTokens.ts#L122-L124.

It is possible that either TS replies with different semantic tokens based on the ranges that we pass in or that we have some sort of bug in the `SparseTokensStore` when combining their reply.

**Comment by AubreyG24** [comment]
Hello, I have never posted a bug issues before, but I am having the same problem with all my files. I have uninstalled VScode and reinstalled it. I got rid of all my extensions and I still have this issues. You can see on the minimap that the code is not highlighting like normal. My JSON file is cleared just has the font-size, theme, and word-wrap on. I attached a video, so I hope that is helpful. Let me know if you have any questions. I also put the Output to hopefully help.  
https://github.com/user-attachments/assets/9e02c7f0-2220-4600-8ed2-8771fafe839f

**Comment by RedCMD** [comment]
@AubreyG24 thats a separate but similar issue
it looks like the initial TextMate highlighting parse isn't happening
and only the viewport parsing is running

I think I've seen that a few times before

prob should make a new issue on it
does it also happen in VSCode web? https://vscode.dev/ and/or https://insiders.vscode.dev/

**Comment by DanteMarshal** [comment]
I'm having a similar issue on C# code. At first I thought it's just the Folding behavior that causes this problem but after seeing @RedCMD 's posts I tried scrolling and realized that I also have this problem when scrolling.

https://github.com/user-attachments/assets/ba167ff8-4d3d-48f7-92dc-271032a7f235

Any updates on this ? Is there any way we can provide more info to help with fixing it ?

**Comment by MichalKrakow** [comment]
Had the same issue with vuejs files, found something called: vue.server.hybridMode in extension config file, switching it off resolved issue for me.

**Comment by Arul-Pando** [comment]
I encountered similar issues with my Vue files, and setting `"vue.server.hybridMode": false` resolved it for me. Thanks @MichalKrakow

**Comment by karsvaniersel** [comment]
> I encountered similar issues with my Vue files, and setting `"vue.server.hybridMode": false` resolved it for me. Thanks [@MichalKrakow](https://github.com/MichalKrakow)

This is the way! I tried many things, but after adding this simple line to my `settings.json` the issue is resolved.

I'm so thankful for your comment, cause this issue was driving me completely insane

**Comment by DanteMarshal** [comment]
> > I encountered similar issues with my Vue files, and setting `"vue.server.hybridMode": false` resolved it for me. Thanks [@MichalKrakow](https://github.com/MichalKrakow)
> 
> This is the way! I tried many things, but after adding this simple line to my `settings.json` the issue is resolved.
> 
> I'm so thankful for your comment, cause this issue was driving me completely insane

This is a Vue-specific workaround and not even a solution. It disables a feature that uses the part of VSCode that's causing this issues as other people including me are encountering this issue in other languages that have nothing to do with Vue or Javascript.

**Comment by RedCMD** [comment]
happening again

https://github.com/user-attachments/assets/2ec2e8ed-6355-4a9f-9281-155010d8e04d

also notice the small delay between scrolling and the colour changing

I reckon it must be caused by the folding ranges
https://github.com/microsoft/vscode/issues/266597

but I don't see any broken ranges

other than missing folding ranges on `new`
```ts
new vscode.Range(
	minifyRange.start,
	minifyRange.start.translate(0, 1),
),
```
but I think that's a different TS bug
<img width="601" height="159" alt="Image" src="https://github.com/user-attachments/assets/ef87885a-e44f-4f5c-a687-042991170cd1" />
